src/python/grpcio_tests/tests/unit/_metadata_flags_test.py

```diff
@@ -0,0 +1,210 @@
+    ('stream_unary', _STREAM_UNARY, True, 'with_call', False),
+    ('stream_unary', _STREAM_UNARY, True, 'future', False),
+    ('stream_stream', _STREAM_STREAM, True, '__call__', True),
+]
+
+
+def handle_unary_unary(test, request, servicer_context):
+    return _RESPONSE
+
+
+def handle_unary_stream(test, request, servicer_context):
+    for _ in range(test_constants.STREAM_LENGTH):
+        yield _RESPONSE
+
+
+def handle_stream_unary(test, request_iterator, servicer_context):
+    for _ in request_iterator:
+        pass
+    return _RESPONSE
+
+
+def handle_stream_stream(test, request_iterator, servicer_context):
+    for _ in request_iterator:
+        yield _RESPONSE
+
+
+class _MethodHandler(grpc.RpcMethodHandler):
+
+    def __init__(self, test, request_streaming, response_streaming):
+        self.request_streaming = request_streaming
+        self.response_streaming = response_streaming
+        self.request_deserializer = None
+        self.response_serializer = None
+        self.unary_unary = None
+        self.unary_stream = None
+        self.stream_unary = None
+        self.stream_stream = None
+        if self.request_streaming and self.response_streaming:
+            self.stream_stream = lambda req, ctx: handle_stream_stream(test, req, ctx)
+        elif self.request_streaming:
+            self.stream_unary = lambda req, ctx: handle_stream_unary(test, req, ctx)
+        elif self.response_streaming:
+            self.unary_stream = lambda req, ctx: handle_unary_stream(test, req, ctx)
+        else:
+            self.unary_unary = lambda req, ctx: handle_unary_unary(test, req, ctx)
+
+
+class _GenericHandler(grpc.GenericRpcHandler):
+
+    def __init__(self, test):
+        self._test = test
+
+    def service(self, handler_call_details):
+        if handler_call_details.method == _UNARY_UNARY:
+            return _MethodHandler(self._test, False, False)
+        elif handler_call_details.method == _UNARY_STREAM:
+            return _MethodHandler(self._test, False, True)
+        elif handler_call_details.method == _STREAM_UNARY:
+            return _MethodHandler(self._test, True, False)
+        elif handler_call_details.method == _STREAM_STREAM:
+            return _MethodHandler(self._test, True, True)
+        else:
+            return None
+
+
+def get_free_loopback_tcp_port():
+    tcp = socket.socket(socket.AF_INET6)
+    tcp.bind(('', 0))
+    address_tuple = tcp.getsockname()
+    return "[%s]:%s" % (address_tuple[0], address_tuple[1])
+
+
+def create_dummy_channel():
+    """Creating dummy channels is a walkaround for retries"""
+    return grpc.insecure_channel(get_free_loopback_tcp_port())
+
+
+def execute_call(channel,
+                 method_name,
+                 server_method,
+                 is_stream_input,
+                 invoke_method,
+                 is_stream_output,
+                 wait_for_ready=None):
+    multi_callable = getattr(channel, method_name)(server_method)
```

Comment:
this way of making an RPC seems fairly complicated, and is more cumbersome than what's shown in the `### DEMO snippets ` section of the gRFC.

Can we change this test (or keep this one but add a new test) to exercise the passing of the `wait_for_ready` param through a stub rather than by using this "stub-bypassed" interface ? Maybe by generating a protobuf for this test? I think that would validate the examples in the gRFC, and also make this test a little simpler

Reply:
This is a hacky way of invocation all eight types of Calls.

1. UnaryUnaryMultiCallable.__call__
1. UnaryUnaryMultiCallable.with_call
1. UnaryUnaryMultiCallable.future
1. UnaryStreamMultiCallable.__call__
1. StreamUnaryMultiCallable.__call__
1. StreamUnaryMultiCallable.with_call
1. StreamUnaryMultiCallable.future
1. StreamStreamMultiCallable.__call__

And the reason why they are complicated is because when input or output is streaming it requires different handling. Once I had a version expanding the almost same code eight times and later decided to shrink it to a single genric function...

Reply:
Personally I'd err on the side of expanding this out, even if there is duplication - it would be nice for readability to be able to compare the described API from the gRFC and the test side by side

Reply:
That's a good point. I will expand them 👍